File: Leetcode-one-by-one/src/Split Array Largest Sum.go
func splitArray(nums []int, k int) int {
	var left, right int64 = 0, 0

	for _, x := range nums {
		if int64(x) > left {
			left = int64(x)
		}
		right += int64(x)
	}

	canSplit := func(limit int64) bool {
		parts := 1
		var curr int64 = 0

		for _, x := range nums {
			if curr+int64(x) > limit {
				parts++
				curr = int64(x)
			} else {
				curr += int64(x)
			}
		}

		return parts <= k
	}

	for left < right {
		mid := left + (right-left)/2

		if canSplit(mid) {
			right = mid
		} else {
			left = mid + 1
		}
	}

	return int(left)
}

File: Leetcode-one-by-one/src/Split Array Largest sum.java
class Solution {
    public int splitArray(int[] nums, int k) {
        long left = 0, right = 0;

        for (int x : nums) {
            left = Math.max(left, x);
            right += x;
        }

        while (left < right) {
            long mid = left + (right - left) / 2;

            if (canSplit(nums, k, mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return (int) left;
    }

    private boolean canSplit(int[] nums, int k, long limit) {
        int parts = 1;
        long curr = 0;

        for (int x : nums) {
            if (curr + x > limit) {
                parts++;
                curr = x;
            } else {
                curr += x;
            }
        }

        return parts <= k;
    }
}
